cpp/src/parquet/arrow/reader_internal.cc

```diff
@@ -612,5 +1121,120 @@ Status TransferColumnData(internal::RecordReader* reader,
   return Status::OK();
 }
 
+Status ReconstructNestedList(const std::shared_ptr<Array>& arr,
+                             std::shared_ptr<Field> field, int16_t max_def_level,
+                             int16_t max_rep_level, const int16_t* def_levels,
+                             const int16_t* rep_levels, int64_t total_levels,
+                             ::arrow::MemoryPool* pool, std::shared_ptr<Array>* out) {
+  if (field->type()->num_children() > 0 && arr->type_id() == ::arrow::Type::DICTIONARY) {
+    // XXX(wesm): Handling of nested types and dictionary encoding needs to be
+    // significantly refactored
+    return Status::Invalid("Cannot have nested types containing dictionary arrays yet");
+  }
+
+  // Walk downwards to extract nullability
+  std::vector<bool> nullable;
+  std::vector<std::shared_ptr<::arrow::Int32Builder>> offset_builders;
```

Comment:
Might be easier than offset_builders and valid_bits_builders:
```
MakeBuilder(pool, list(list(list(/*...*/null()/*...*/))), &root_builder);
```

Reply:
Old code